Add support for UEFI

pirat89 opened the pull request:
## Brief description
After the conversion to RHEL system on UEFI, usually these errors are
present:
 1) Missing grub files (grubenv, grub.cfg, ..) insinde the
    /boot/efi/EFI/redhat directory [CentOS Linux only]
 2) Current (default) bootloader points to non-existing files
    [CentOS Linux only]
 3) New bootloader entry for RHEL doesn't exist
    (in case the original entry points to the efi redhat directory,
     just the label is missleading, speaking about the orig OS)
  4) similar to 1). in case of non-standard EFI setup, user could
      put stuff into the custom directory. In such a case, bootloader setup
      (grubenv, grub.cfg) copied into the redhat directory will not contain
      the expected grub entries - with pointer to RHEL kernel. As well,
      grub.cfg could be renamed to something what address the efi bin
      filename (if I remember well). 

For CentOS, the first two issues are critical and lead into the broken
bootloader. For OL this is not blocking as the redhat efi dir is used
already (paths are same by default). This commit is resolving these
problems, as it enshurs that all expected files are present inside
the redhat efi dir and creates a new EFI bootloader entry pointing
to the default EFI binary file:
  `\\EFI\\redhat\\shimx64.bin` or `\\EFI\\redhat\\grubx64.bin`
The case 4) is kind of non-standard but it could happen. It's relevant
only for a non-secure boot (the secure boot has to follow standard
very strictly).

---
## specification
Here is list to describe the covered cases and limitations:
- covered default setups for OL & Centos - including with/without shim installed
  - expected now at least use of standard path to a directory with a used EFI binary file
    - if this condition is not met, which is non-standard, consider it as a missing implementation in this PR now (see TODO)
  - for centos, grub.cfg, grubenv, and user.cfg files are copied into expected redhat directory after the conversion
  -  new EFI bootloader entry referring to the installed "Red Hat Enterprise Linux X" system is created
      - shimx64.efi from RHEL is used as the default EFI bin bootloader if installed, otherwise fallback to grubx64.efi (must be installed!)
  - currently the current EFI bootloader entry is removed, unless:
    - the reffered EFI binary file still exists
    - cannot get canonical path to the EFI bin (unexpected; it can happen if booted via PXE e.g.)
    - the entry has been modified already during the  bootloader migration (unexpected; it can happen only in case of bug in the `convert2rhel.grub` module)
- added inhibitors (when EFI detected) for
  - OL/CentOS 6
  - non-intel architectures
  - secure-boot - this one is untested. potentially it will works, but in case user use a custom stuff (e.g. self signed stuff)
  - systems without efibootmgr (this is kinda unsupported)
  - systems where we are not able to obtain all expected information about the bootloader configuration (e.g. because of bug or totally misconfigured systems - we know already some examples)
- if multiple EFI bootloader entries for CentOS exists, some of them could survived untouched. E.g. point to a bootloader with non-existing kernel entries, speak about CentOS Linux when it's not longer existing, etc. I've kept that for now, as I can imagine user could create manually a custom bootloader, pointing to an LVM snapshot where the original system could still exists) and which could be working actually.
- in case that any post PONR step of the bootloader migration fails, the critical message is logged with information the bootloader has to be migrated manually before the reboot.

------
## Tips for manual testing.
If you want to test various UEFI setups in your VM, it could be benefitial for you to use snapshots. In case of UEFI, the external snapshots can be used only. However virsh does not provide nice CLI support for that. To make it more simple, I have created a bash library  based on `virsh` to manipulate with snaphots easier. See: https://github.com/pirat89/mydots/blob/master/bash_libs/bashvirt_ext  - pay attention to `_LIBVIRT_XML_BACKUP_DIR` and `_DO_NOT_PROMPT_REMOVAL` variables (see the comments). The three methods are relevant:
```
virt_snap_ext
virt_snap_ext_reload
virt_snap_ext_delete_last
```

When these are called without arguments, it will print usage.. As well, you can read comments inside the library. The library expects you have configured your environment correctly already to be able to use virsh without the root permissions. For the setup, do something like this:
```bash
echo  >> /etc/libvirt/libvirtd.conf 'unix_sock_group = "libvirt"' 
echo  >> /etc/libvirt/libvirtd.conf 'unix_sock_ro_perms = "0777"'
echo  >> /etc/libvirt/libvirtd.conf 'unix_sock_rw_perms = "0770"'
echo  >> /etc/libvirt/libvirtd.conf 'auth_unix_ro = "none"'
echo  >> /etc/libvirt/libvirtd.conf 'auth_unix_rw = "none"'
usermod -aG libvirt <username>
```
reminder - if you have not been in the libvirt group before, you will need to restart your session too.

----

TODO:
 - [x] use grubx64.efi in case the shim package is not installed
       (or handle the problem in a different way) (going to be discussed with the bootloader team)
       - update: there is not problem to use e.g. grubx64.efi file when the shim is not installed, unless secure boot is used
 - [x] ~What to do with other ORIG_OS entries EFI bootloaders?
       Additional centos / OL bootloaders could contain missleading
       labels speaking about original OS which is not present anymore.
       However, if points e.g. to an LVM snapshot with centos..?
       Do we need to deal with this?~ (ignore: We will deal with it once people ask about that)
 - [x] ~Remove the original efi centos directory when no other efi
       bins are present~ (ignore: currently it's planned to keep the directory in this PR; can be resolved in future)
 - [x] Fix the error handling
 - [ ] What about systems using completely different directory for
       the default bootloader? (ignore: we will always go to our dir & efi file)
       - UPDATE: It's valid. This could be potentially problem as a different grub files can be used (so our grub entry for the rhel kernel doesn't have to exist)
- [x] Detect secure boot (I vote for the inhibitior now and do additional testing with allowed secure boot later)
- [x] Do not remove the original EFI bootloader if the target file still exists
- [x] Make the ERROR log colored (red)

Last TODO:
- [x] Unit tests for the checks module
- [ ] Unit tests for the grub module

pirat89 commented:
I am going to resolve TODO items basd on discussion with Michal. Once I am finished, I will sent the review request. I will start to implement unit-tests after all other stuff is finished.

shaded-enmity commented on convert2rhel/grub.py:
```diff
@@ -0,0 +1,493 @@
+    Raise UnsupportedEFIConfiguration when ESP is not mounted where expected.
+    Raise BootloaderError if the partition cannot be obtained from GRUB.
+    """
+    if not is_uefi:
```
```suggestion
    if not is_uefi():
```

pirat89 replied:
Fixed (with the next push).

pirat89 commented:
I haven't tested the code yet, however this is basically it. I have realized just now  one potential problem with the grub files in case the current efi boot is non default / points to an own directory, which I need to test as well as it could be fatal. However, unless I did mistake (I bet I will discover bunch of bugs when I test it), this should be it.

I will do the manual testing tomorrow and fix all discovered errors. After that I will start to work on the unit tests. Basically it's prepared for the pre-review. Once I finish my own testing and fixes, I will let you know.

pirat89 commented:
The code is prepared for pre-review. This is basically it

shaded-enmity reviewed (commented):
A couple of comments :)

I'd also suggest picking either EFI or UEFI and use the term consistently, right now both code/logs use one or the other in kinda random fashion.

pirat89 commented on convert2rhel/grub.py:
```diff
@@ -0,0 +1,597 @@
+
+def is_secure_boot():
+    """Return True if the secure boot is enabled."""
+    if not is_uefi:
```
oh my... this is my favourite typo I guess :dancing_women:

pirat89 commented on convert2rhel/checks.py:
```diff
@@ -40,17 +42,50 @@ def perform_pre_checks():
+    try:
+        efiboot_info = grub.EFIBootInfo()
+    except grub.BootloaderError as e:
+        # NOTE(pstodulk): maybe just print of the error msg could be enough..
```
@shaded-enmity sorry, wrong wording :D The NOTE is that probably this could be enough:
```
logger.critical(e.message)
```

it always counts with the critical level -> inhibitor. WDYT? should I keep the msg as it is or just print the error message from the exception? I can remove the note then.

shaded-enmity replied:
ahh, makes sense. Yeah I'd just print `e.message` to make it more straightforward.

pirat89 replied:
Done.

pirat89 commented:
Rebased against the current master and squashed all UEFI commits. In case someone would want to see more fragmented development the original commits are backed up in the [uefi-enable-backup](https://github.com/pirat89/convert2rhel/tree/uefi-enable-backup) branch. Regarding the upstream development is fast and it's not sure when this will be finished (`grub.py` library is not covered by unit-tests and conversion tests are missing) I want to reduce possible problems in future with rebases.

bocekm commented:
Thanks for the rebase, @pirat89.

ZhukovGreen commented on convert2rhel/unit_tests/checks_test.py:
```diff
@@ -473,31 +457,153 @@ def test_check_tainted_kmods(monkeypatch, command_return, expected_exception):
+        checks.check_tainted_kmods()
+
+
+class EFIBootInfoMocked():
```
The interface of this object is completely detached from the original one. Is it possible to inherit from the origin one and just override some functionality?

ZhukovGreen replied:
maybe Mock(autospec=True)?

pirat89 replied:
What is the original one? Can you elaborate on it?

ZhukovGreen replied:
The original one is the one we mock - class  EfiBootInfo (https://github.com/oamg/convert2rhel/pull/223/files#diff-3a7eda59c836f39329b256971a1603cf9345a87f7abc30d6c681b5da961967b1R258). So using for example autospeccing (https://docs.python.org/3/library/unittest.mock.html#autospeccing) might improve the quality of the test and make it more useful. I think what we can do here is something which shown at the end of the chapter:
```python
>>> class Something:
...   def __init__(self):
...     self.a = 33
...
>>> class SomethingForTest(Something):
...   a = 33
...
>>> p = patch('__main__.Something', autospec=SomethingForTest)
>>> mock = p.start()
>>> mock.a
<NonCallableMagicMock name='Something.a' spec='int' id='...'>
```

ZhukovGreen reviewed (commented):
@pirat89 some surface review provided. It does not cover all, as I expect there will be changes

pirat89 commented on convert2rhel/grub.py:
```diff
@@ -0,0 +1,597 @@
+            # it's not expected that no entry exists
+            raise BootloaderError("EFI: Cannot detect EFI bootloaders.")
+
+    def _parse_current_boot(self, data):
```
That's probably good idea

ZhukovGreen replied:
leaving it up to you

pirat89 commented on convert2rhel/grub.py:
```diff
@@ -0,0 +1,597 @@
+    def _parse_efi_boot_entries(self, brief_data, verbose_data):
+        """Return dict of EFI boot loader entries: {"<boot_number>": EFIBootLoader}"""
+        self.entries = {}
+        regexp_entry = re.compile(r"^Boot(?P<bootnum>[0-9]+)[\s*]\s*(?P<label>[^\s].*)$")
```
Doesn't matter when you do not expose the regexp in the module (which was my intention). We are literally talking about zero saved time. Do not think about optimisation when it doesn't make sense. This function you want to call just couple of times during the whole conversion.

pirat89 commented on convert2rhel/grub.py:
```diff
@@ -0,0 +1,597 @@
+        """Return dict of EFI boot loader entries: {"<boot_number>": EFIBootLoader}"""
+        self.entries = {}
+        regexp_entry = re.compile(r"^Boot(?P<bootnum>[0-9]+)[\s*]\s*(?P<label>[^\s].*)$")
+        for line in brief_data.splitlines():
+            match = regexp_entry.match(line)
+            if not match:
+                continue
```
This is quite opposite. The original solution is much more simple & readable.

pirat89 commented on convert2rhel/grub.py:
```diff
@@ -0,0 +1,597 @@
+    if ecode:
+        logger.error("Cannot get information about the '%s' device." % device)
+        logger.debug("lsblk ... output:\n-----\n%s\n-----" % stdout)
+        return None
```
You are right. The docstring should be corrected.

pirat89 replied:
Ah. Now I get it. I meant something different. forget it.

pirat89 commented on convert2rhel/grub.py:
```diff
@@ -0,0 +1,597 @@
+        stdout, ecode = utils.run_subprocess("mokutil --sb-state", print_output=False)
+    except OSError:
+        return False
+    if ecode or "enabled" not in stdout:
```
oki

pirat89 commented on convert2rhel/grub.py:
```diff
@@ -0,0 +1,597 @@
+
+    Raise BootloaderError if the partition cannot be detected.
+    """
+    return _get_partition("/boot")
```
For me it makes sense, as from the module is expected you want to get only the boot partition of the efi partition. Nothing else. The module should not provide a function for a whatever parition and it doesn't make sense to write the code multiple times.

ZhukovGreen reviewed (commented):
@pirat89 some answers. Do not take it as something that has to be done in order to merge. I can help you with the implementation if we agree that my particular suggestion has sense. Thanks for helping us with this feature

allquixotic commented:
Hi,

On the OVH image of CentOS 8, I tried this branch and received the following result with no customizations to the OS other than installing the required python3 dependencies for convert2rhel.

```
[05/06/2021 20:59:34] TASK - [Convert: Configure the bootloader] ********************************
The new expected EFI binary found: /boot/efi/EFI/redhat/grubx64.efi
Copy the GRUB2 configuration files to the new EFI directory.
ERROR - Cannot find the original file required for the proper configuration: /boot/efi/EFI/centos/grubenv
Copying '/boot/efi/EFI/centos/grub.cfg' to '/boot/efi/EFI/redhat/grub.cfg'
CRITICAL - Some GRUB files have not been copied to /boot/efi/EFI/redhat
The migration of the bootloader setup was not successful.
Do not reboot your machine before the manual check of the
bootloader configuration. Ensure grubenv and grub.cfg files
are present inside the /boot/efi/EFI/redhat/ directory and
the new bootloader entry for Red Hat Enterprise Linux exist
(check `efibootmgr -v` output).
The entry should point to '\EFI\redhat\shimx64.efi'.
WARNING - Conversion process interrupted and manual user intervention will be necessary.
```

This may be useful to you for debugging:

```
[root@tres convert2rhel]# find /boot
/boot
/boot/efi
/boot/efi/EFI
/boot/efi/EFI/centos
/boot/efi/EFI/centos/grub.cfg.rpmsave
/boot/efi/EFI/centos/grub.cfg
/boot/efi/EFI/BOOT
/boot/efi/EFI/redhat
/boot/efi/EFI/redhat/fonts
/boot/efi/EFI/redhat/grub.cfg
/boot/efi/EFI/redhat/grubx64.efi
/boot/grub
/boot/grub/menu.lst
/boot/grub/grub.conf
/boot/grub2
/boot/grub2/fonts
/boot/grub2/fonts/unicode.pf2
/boot/grub2/i386-pc
/boot/grub2/i386-pc/fat.mod
/boot/grub2/i386-pc/acpi.mod
/boot/grub2/i386-pc/file.mod
/boot/grub2/i386-pc/adler32.mod
/boot/grub2/i386-pc/gcry_md5.mod
/boot/grub2/i386-pc/affs.mod
/boot/grub2/i386-pc/gcry_rfc2268.mod
/boot/grub2/i386-pc/afs.mod
/boot/grub2/i386-pc/gcry_rijndael.mod
/boot/grub2/i386-pc/ahci.mod
/boot/grub2/i386-pc/gcry_rmd160.mod
/boot/grub2/i386-pc/all_video.mod
/boot/grub2/i386-pc/gcry_rsa.mod
/boot/grub2/i386-pc/aout.mod
/boot/grub2/i386-pc/gcry_seed.mod
/boot/grub2/i386-pc/archelp.mod
/boot/grub2/i386-pc/eval.mod
/boot/grub2/i386-pc/at_keyboard.mod
/boot/grub2/i386-pc/gcry_serpent.mod
/boot/grub2/i386-pc/ata.mod
/boot/grub2/i386-pc/gcry_sha1.mod
/boot/grub2/i386-pc/backtrace.mod
/boot/grub2/i386-pc/gcry_sha256.mod
/boot/grub2/i386-pc/bfs.mod
/boot/grub2/i386-pc/gcry_sha512.mod
/boot/grub2/i386-pc/biosdisk.mod
/boot/grub2/i386-pc/gcry_tiger.mod
/boot/grub2/i386-pc/bitmap.mod
/boot/grub2/i386-pc/exfat.mod
/boot/grub2/i386-pc/bitmap_scale.mod
/boot/grub2/i386-pc/gcry_twofish.mod
/boot/grub2/i386-pc/blocklist.mod
/boot/grub2/i386-pc/gcry_whirlpool.mod
/boot/grub2/i386-pc/blscfg.mod
/boot/grub2/i386-pc/gdb.mod
/boot/grub2/i386-pc/boot.img
/boot/grub2/i386-pc/geli.mod
/boot/grub2/i386-pc/boot.mod
/boot/grub2/i386-pc/gettext.mod
/boot/grub2/i386-pc/bsd.mod
/boot/grub2/i386-pc/exfctest.mod
/boot/grub2/i386-pc/bswap_test.mod
/boot/grub2/i386-pc/gfxmenu.mod
/boot/grub2/i386-pc/btrfs.mod
/boot/grub2/i386-pc/gfxterm.mod
/boot/grub2/i386-pc/bufio.mod
/boot/grub2/i386-pc/gfxterm_menu.mod
/boot/grub2/i386-pc/cat.mod
/boot/grub2/i386-pc/gzio.mod
/boot/grub2/i386-pc/cbfs.mod
/boot/grub2/i386-pc/gptsync.mod
/boot/grub2/i386-pc/cbls.mod
/boot/grub2/i386-pc/halt.mod
/boot/grub2/i386-pc/cbmemc.mod
/boot/grub2/i386-pc/hashsum.mod
/boot/grub2/i386-pc/cbtable.mod
/boot/grub2/i386-pc/hdparm.mod
/boot/grub2/i386-pc/cbtime.mod
/boot/grub2/i386-pc/hello.mod
/boot/grub2/i386-pc/chain.mod
/boot/grub2/i386-pc/gcry_des.mod
/boot/grub2/i386-pc/cmdline_cat_test.mod
/boot/grub2/i386-pc/help.mod
/boot/grub2/i386-pc/cmosdump.mod
/boot/grub2/i386-pc/hexdump.mod
/boot/grub2/i386-pc/cmostest.mod
/boot/grub2/i386-pc/hfs.mod
/boot/grub2/i386-pc/cmp.mod
/boot/grub2/i386-pc/hfsplus.mod
/boot/grub2/i386-pc/cmp_test.mod
/boot/grub2/i386-pc/increment.mod
/boot/grub2/i386-pc/command.lst
/boot/grub2/i386-pc/ext2.mod
/boot/grub2/i386-pc/configfile.mod
/boot/grub2/i386-pc/hfspluscomp.mod
/boot/grub2/i386-pc/core.img
/boot/grub2/i386-pc/http.mod
/boot/grub2/i386-pc/cpio.mod
/boot/grub2/i386-pc/iorw.mod
/boot/grub2/i386-pc/cpio_be.mod
/boot/grub2/i386-pc/iso9660.mod
/boot/grub2/i386-pc/cpuid.mod
/boot/grub2/i386-pc/jfs.mod
/boot/grub2/i386-pc/crc64.mod
/boot/grub2/i386-pc/jpeg.mod
/boot/grub2/i386-pc/crypto.lst
/boot/grub2/i386-pc/ldm.mod
/boot/grub2/i386-pc/crypto.mod
/boot/grub2/i386-pc/extcmd.mod
/boot/grub2/i386-pc/cryptodisk.mod
/boot/grub2/i386-pc/keylayouts.mod
/boot/grub2/i386-pc/cs5536.mod
/boot/grub2/i386-pc/keystatus.mod
/boot/grub2/i386-pc/ctz_test.mod
/boot/grub2/i386-pc/legacycfg.mod
/boot/grub2/i386-pc/date.mod
/boot/grub2/i386-pc/linux.mod
/boot/grub2/i386-pc/datehook.mod
/boot/grub2/i386-pc/loadenv.mod
/boot/grub2/i386-pc/datetime.mod
/boot/grub2/i386-pc/loopback.mod
/boot/grub2/i386-pc/disk.mod
/boot/grub2/i386-pc/f2fs.mod
/boot/grub2/i386-pc/diskfilter.mod
/boot/grub2/i386-pc/ls.mod
/boot/grub2/i386-pc/div.mod
/boot/grub2/i386-pc/lsacpi.mod
/boot/grub2/i386-pc/div_test.mod
/boot/grub2/i386-pc/lsapm.mod
/boot/grub2/i386-pc/dm_nv.mod
/boot/grub2/i386-pc/lsmmap.mod
/boot/grub2/i386-pc/drivemap.mod
/boot/grub2/i386-pc/lspci.mod
/boot/grub2/i386-pc/echo.mod
/boot/grub2/i386-pc/luks.mod
/boot/grub2/i386-pc/efiemu.mod
/boot/grub2/i386-pc/lvm.mod
/boot/grub2/i386-pc/ehci.mod
/boot/grub2/i386-pc/lzopio.mod
/boot/grub2/i386-pc/elf.mod
/boot/grub2/i386-pc/font.mod
/boot/grub2/i386-pc/macbless.mod
/boot/grub2/i386-pc/freedos.mod
/boot/grub2/i386-pc/macho.mod
/boot/grub2/i386-pc/fs.lst
/boot/grub2/i386-pc/mdraid09_be.mod
/boot/grub2/i386-pc/fshelp.mod
/boot/grub2/i386-pc/gcry_crc.mod
/boot/grub2/i386-pc/functional_test.mod
/boot/grub2/i386-pc/gcry_dsa.mod
/boot/grub2/i386-pc/gcry_arcfour.mod
/boot/grub2/i386-pc/gfxterm_background.mod
/boot/grub2/i386-pc/gcry_blowfish.mod
/boot/grub2/i386-pc/gcry_idea.mod
/boot/grub2/i386-pc/gcry_camellia.mod
/boot/grub2/i386-pc/gcry_md4.mod
/boot/grub2/i386-pc/gcry_cast5.mod
/boot/grub2/i386-pc/mda_text.mod
/boot/grub2/i386-pc/relocator.mod
/boot/grub2/i386-pc/legacy_password_test.mod
/boot/grub2/i386-pc/mdraid09.mod
/boot/grub2/i386-pc/setpci.mod
/boot/grub2/i386-pc/mdraid1x.mod
/boot/grub2/i386-pc/sfs.mod
/boot/grub2/i386-pc/memdisk.mod
/boot/grub2/i386-pc/shift_test.mod
/boot/grub2/i386-pc/memrw.mod
/boot/grub2/i386-pc/signature_test.mod
/boot/grub2/i386-pc/minicmd.mod
/boot/grub2/i386-pc/sleep.mod
/boot/grub2/i386-pc/minix.mod
/boot/grub2/i386-pc/sleep_test.mod
/boot/grub2/i386-pc/minix2.mod
/boot/grub2/i386-pc/spkmodem.mod
/boot/grub2/i386-pc/minix2_be.mod
/boot/grub2/i386-pc/squash4.mod
/boot/grub2/i386-pc/minix3.mod
/boot/grub2/i386-pc/strtoull_test.mod
/boot/grub2/i386-pc/minix3_be.mod
/boot/grub2/i386-pc/syslinuxcfg.mod
/boot/grub2/i386-pc/minix_be.mod
/boot/grub2/i386-pc/tar.mod
/boot/grub2/i386-pc/mmap.mod
/boot/grub2/i386-pc/terminal.lst
/boot/grub2/i386-pc/moddep.lst
/boot/grub2/i386-pc/terminal.mod
/boot/grub2/i386-pc/modinfo.sh
/boot/grub2/i386-pc/terminfo.mod
/boot/grub2/i386-pc/morse.mod
/boot/grub2/i386-pc/test.mod
/boot/grub2/i386-pc/mpi.mod
/boot/grub2/i386-pc/test_blockarg.mod
/boot/grub2/i386-pc/msdospart.mod
/boot/grub2/i386-pc/testload.mod
/boot/grub2/i386-pc/mul_test.mod
/boot/grub2/i386-pc/testspeed.mod
/boot/grub2/i386-pc/multiboot.mod
/boot/grub2/i386-pc/romfs.mod
/boot/grub2/i386-pc/multiboot2.mod
/boot/grub2/i386-pc/scsi.mod
/boot/grub2/i386-pc/nativedisk.mod
/boot/grub2/i386-pc/tftp.mod
/boot/grub2/i386-pc/net.mod
/boot/grub2/i386-pc/tga.mod
/boot/grub2/i386-pc/newc.mod
/boot/grub2/i386-pc/time.mod
/boot/grub2/i386-pc/nilfs2.mod
/boot/grub2/i386-pc/tr.mod
/boot/grub2/i386-pc/normal.mod
/boot/grub2/i386-pc/trig.mod
/boot/grub2/i386-pc/ntfs.mod
/boot/grub2/i386-pc/true.mod
/boot/grub2/i386-pc/ntfscomp.mod
/boot/grub2/i386-pc/truecrypt.mod
/boot/grub2/i386-pc/ntldr.mod
/boot/grub2/i386-pc/udf.mod
/boot/grub2/i386-pc/odc.mod
/boot/grub2/i386-pc/ufs1.mod
/boot/grub2/i386-pc/offsetio.mod
/boot/grub2/i386-pc/ufs1_be.mod
/boot/grub2/i386-pc/ohci.mod
/boot/grub2/i386-pc/search.mod
/boot/grub2/i386-pc/part_acorn.mod
/boot/grub2/i386-pc/search_fs_file.mod
/boot/grub2/i386-pc/part_amiga.mod
/boot/grub2/i386-pc/search_fs_uuid.mod
/boot/grub2/i386-pc/part_apple.mod
/boot/grub2/i386-pc/ufs2.mod
/boot/grub2/i386-pc/part_bsd.mod
/boot/grub2/i386-pc/uhci.mod
/boot/grub2/i386-pc/part_dfly.mod
/boot/grub2/i386-pc/usb.mod
/boot/grub2/i386-pc/part_dvh.mod
/boot/grub2/i386-pc/usb_keyboard.mod
/boot/grub2/i386-pc/part_gpt.mod
/boot/grub2/i386-pc/search_label.mod
/boot/grub2/i386-pc/part_msdos.mod
/boot/grub2/i386-pc/usbms.mod
/boot/grub2/i386-pc/part_plan.mod
/boot/grub2/i386-pc/usbserial_common.mod
/boot/grub2/i386-pc/part_sun.mod
/boot/grub2/i386-pc/sendkey.mod
/boot/grub2/i386-pc/part_sunpc.mod
/boot/grub2/i386-pc/usbserial_ftdi.mod
/boot/grub2/i386-pc/partmap.lst
/boot/grub2/i386-pc/usbserial_pl2303.mod
/boot/grub2/i386-pc/parttool.lst
/boot/grub2/i386-pc/usbtest.mod
/boot/grub2/i386-pc/parttool.mod
/boot/grub2/i386-pc/vbe.mod
/boot/grub2/i386-pc/password.mod
/boot/grub2/i386-pc/serial.mod
/boot/grub2/i386-pc/password_pbkdf2.mod
/boot/grub2/i386-pc/verify.mod
/boot/grub2/i386-pc/pata.mod
/boot/grub2/i386-pc/vga.mod
/boot/grub2/i386-pc/pbkdf2.mod
/boot/grub2/i386-pc/setjmp.mod
/boot/grub2/i386-pc/pbkdf2_test.mod
/boot/grub2/i386-pc/vga_text.mod
/boot/grub2/i386-pc/pci.mod
/boot/grub2/i386-pc/video.lst
/boot/grub2/i386-pc/pcidump.mod
/boot/grub2/i386-pc/video.mod
/boot/grub2/i386-pc/plan9.mod
/boot/grub2/i386-pc/video_bochs.mod
/boot/grub2/i386-pc/play.mod
/boot/grub2/i386-pc/video_cirrus.mod
/boot/grub2/i386-pc/png.mod
/boot/grub2/i386-pc/setjmp_test.mod
/boot/grub2/i386-pc/priority_queue.mod
/boot/grub2/i386-pc/video_colors.mod
/boot/grub2/i386-pc/probe.mod
/boot/grub2/i386-pc/video_fb.mod
/boot/grub2/i386-pc/procfs.mod
/boot/grub2/i386-pc/videoinfo.mod
/boot/grub2/i386-pc/progress.mod
/boot/grub2/i386-pc/videotest.mod
/boot/grub2/i386-pc/pxe.mod
/boot/grub2/i386-pc/xfs.mod
/boot/grub2/i386-pc/pxechain.mod
/boot/grub2/i386-pc/xnu.mod
/boot/grub2/i386-pc/raid5rec.mod
/boot/grub2/i386-pc/xnu_uuid.mod
/boot/grub2/i386-pc/raid6rec.mod
/boot/grub2/i386-pc/xnu_uuid_test.mod
/boot/grub2/i386-pc/random.mod
/boot/grub2/i386-pc/xzio.mod
/boot/grub2/i386-pc/read.mod
/boot/grub2/i386-pc/zfs.mod
/boot/grub2/i386-pc/reboot.mod
/boot/grub2/i386-pc/zfscrypt.mod
/boot/grub2/i386-pc/regexp.mod
/boot/grub2/i386-pc/zfsinfo.mod
/boot/grub2/i386-pc/reiserfs.mod
/boot/grub2/i386-pc/usbserial_usbdebug.mod
/boot/grub2/i386-pc/videotest_checksum.mod
/boot/grub2/device.map
/boot/grub2/grub.cfg
/boot/grub2/grubenv
/boot/grub2/grubenv.rpmnew
/boot/loader
/boot/loader/entries
/boot/loader/entries/bdefe75b158741fb83a0be46662c6c8e-0-rescue.conf
/boot/loader/entries/bdefe75b158741fb83a0be46662c6c8e-4.18.0-240.15.1.el8_3.x86_64.conf
/boot/loader/entries/bdefe75b158741fb83a0be46662c6c8e-4.18.0-240.22.1.el8_3.x86_64.conf
/boot/initramfs-0-rescue-6bdfa40330b14232b9c78e13592904be.img
/boot/initramfs-0-rescue-bdefe75b158741fb83a0be46662c6c8e.img
/boot/vmlinuz-0-rescue-6bdfa40330b14232b9c78e13592904be
/boot/vmlinuz-0-rescue-bdefe75b158741fb83a0be46662c6c8e
/boot/initramfs-4.18.0-240.10.1.el8_3.x86_64kdump.img
/boot/vmlinuz-4.18.0-240.15.1.el8_3.x86_64
/boot/System.map-4.18.0-240.15.1.el8_3.x86_64
/boot/config-4.18.0-240.15.1.el8_3.x86_64
/boot/.vmlinuz-4.18.0-240.15.1.el8_3.x86_64.hmac
/boot/initramfs-4.18.0-240.15.1.el8_3.x86_64.img
/boot/vmlinuz-4.18.0-240.22.1.el8_3.x86_64
/boot/System.map-4.18.0-240.22.1.el8_3.x86_64
/boot/config-4.18.0-240.22.1.el8_3.x86_64
/boot/.vmlinuz-4.18.0-240.22.1.el8_3.x86_64.hmac
/boot/initramfs-4.18.0-240.22.1.el8_3.x86_64.img
```

Given that OVH is a pretty decent-sized company (see: https://www.ovh.com/world/us/about-us/datacenters.xml ) would it make sense to special-case support for the OVH image in convert2rhel?

pirat89 commented:
@allquixotic thanks for the feedback. I would tell it's kind of unexpected regarding the  `/boot/grub2/grubenv` is not symlink as I'd expect. Definitely good to know the problem exists on OVH Centos images. Can you check what's the `/boot/grub2/grubenv.rpmnew` file? Better, could you share `ls -l  /boot/grub2/grubenv*` for those two files?   Not sure what will be plans about support. @bocekm Michal, that's good question for you.

From the technical POV, we will need to discuss the setup with our bootloader guys whether it's supported / correct on RHEL 8 just for sure (it could be e.g. even wrongly create image; I dunno tbh, just I know already it's better to check it). And based on whether the grubenv file is symlink or not on OVH RHEL, there are possible simple solutions for the fix.

In case that support of OVH way is planned to be skipped at least for now in this first iteration, we should at least create an inhibitor to prevent unexpected failures.

joreiff commented:
Please note that CentOS 8 will go EOL at the end of 2021. It would be awesome if this feature could be merged before that. Thanks for the great work!

SpyTec reviewed (commented):
I'll take over and try to do what's left once I have a grasp of what needs to be done. Absolutely fantastic docstrings and comments in the code and overall code structure looks great.

Most comments are just minor changes. Three of them are questions

SpyTec commented:
/packit build

pirat89 commented on convert2rhel/grub.py:
```diff
@@ -0,0 +1,597 @@
+DEFAULT_INSTALLED_EFIBIN_FILENAMES = ("shimx64.efi", "grubx64.efi")
+"""Filenames of the EFI binary files that could be by default instaled on the system.
+
+Sorted by the recommended preferences. The first one is most preferred, but it
+is provided by the shim rpm which does not have to be installed on the system.
+In case it's missing, another files should be used instead.
```
Discussed in person, but addint the comment for other possible readers :)

In case of the secure boot the `shimx64` is required, otherwise any of these files can be used. But in case a user will want to switch to the secure boot (not during the conversion but whenever in future), it's easier for them to do it when the `shimx64` file is used. So the `shimx64` is prefered to use because of this reason only.

SpyTec commented on convert2rhel/grub.py:
```diff
@@ -0,0 +1,597 @@
+    efibootinfo_new = EFIBootInfo()
+    # the following checks are just preventive. the warnings should not appear,
+    # unless someone changes the code in future providing a bug
+    orig_boot = efibootinfo_new.entries.get(efibootinfo_orig.current_boot, None)
+    if not orig_boot:
+        logger.warning(
+            "The original default EFI bootloader entry %s has been removed already."
+            % orig_boot.boot_number
+        )
+        return
+    if orig_boot != efibootinfo_orig.entries[orig_boot.boot_number]:
+        logger.warning(
+            "The EFI bootloader entry %s has been modified. Skipping the removal."
+            % orig_boot.boot_number
+        )
+        return
```
@pirat89 @bocekm I've setup a CentOS 8 with UEFI and either the VM is setup incorrectly or the creation of a boot entry is incorrect and would want to hear your thoughts.

Before creation we have this. You'll notice there isn't an entry for `0001`, so when we remove CentOS (`0004`) and add in a new entry for RHEL, it gets assigned `0001`. When we then do the check to make sure they're different we make an assumption it will be in the same spot
```
[root@localhost ~]# /usr/sbin/efibootmgr
BootCurrent: 0004
Timeout: 0 seconds
BootOrder: 0004,0002,0000,0003
Boot0000* UiApp
Boot0002* UEFI Misc Device
Boot0003* EFI Internal Shell
Boot0004* CentOS Linux
```

Should we instead keep track of the new bootloader number and compare that instead of assuming it will be the same number as the CentOS one? Or is this handling a case for an unexpected configuration? Or am I completely misunderstanding what is going on? :D

This is the logs with changes to bootloader
```
[...]
[09/09/2021 12:14:45] DEBUG - EFI: the next boot is not set.
Remove the original EFI boot entry '0004'.
[09/09/2021 12:14:45] DEBUG - Calling command '/usr/sbin/efibootmgr -Bb 0004'

[09/09/2021 12:14:45] TASK - [Convert: Patch yum configuration file] ****************************                 
[09/09/2021 12:14:45] DEBUG - Calling command 'rpm -Vf /etc/dnf/dnf.conf'
Skipping patching, yum configuration file not modified

[09/09/2021 12:14:45] TASK - [Final: rpm files modified by the conversion] **********************                 
Skipping the execution of 'rpm -Va'.
Skipping comparison of the 'rpm -Va' output from before and after the conversion.

[09/09/2021 12:14:45] TASK - [Final: Configure the bootloader] **********************************                 
The new expected EFI binary found: /boot/efi/EFI/redhat/shimx64.efi
Copy the GRUB2 configuration files to the new EFI directory.
The /boot/efi/EFI/redhat/grubenv file already exists. Copying skipped.
The /boot/efi/EFI/redhat/grub.cfg file already exists. Copying skipped.
The original /boot/efi/EFI/centos directory is kept. Remove the directory manually after you check it's not needed anymore.
Load the bootloader configuration.
[09/09/2021 12:14:45] DEBUG - Calling command '/usr/sbin/efibootmgr'
BootCurrent: 0004
Timeout: 0 seconds
BootOrder: 0001,0002,0000,0003
Boot0000* UiApp
Boot0001* Red Hat Enterprise Linux 8
Boot0002* UEFI Misc Device
Boot0003* EFI Internal Shell
[09/09/2021 12:14:45] DEBUG - Calling command '/usr/sbin/efibootmgr -v'
BootCurrent: 0004
Timeout: 0 seconds
BootOrder: 0001,0002,0000,0003
Boot0000* UiApp FvVol(7cb8bdc9-f8eb-4f34-aaea-3ee4af6516a1)/FvFile(462caa21-7614-4503-836e-8ab6f4662331)
Boot0001* Red Hat Enterprise Linux 8    HD(1,GPT,714e014a-5636-47b0-a8a7-b5109bfe895c,0x800,0x12c000)/File(\EFI\redhat\shimx64.efi)
Boot0002* UEFI Misc Device      PciRoot(0x0)/Pci(0x2,0x3)/Pci(0x0,0x0)N.....YM....R,Y.
Boot0003* EFI Internal Shell    FvVol(7cb8bdc9-f8eb-4f34-aaea-3ee4af6516a1)/FvFile(7c04a583-9e3e-4f1c-ad65-e05268d0b4d1)
[09/09/2021 12:14:45] DEBUG - Calling command '/usr/sbin/grub2-probe --target=device /boot/efi/'
[09/09/2021 12:14:45] DEBUG - EFI: the next boot is not set.
Replace the current EFI bootloader entry with the RHEL one.
Create the 'Red Hat Enterprise Linux 8' EFI bootloader entry.
[09/09/2021 12:14:45] DEBUG - Calling command 'lsblk -spnlo MAJ:MIN /dev/vda1'
[09/09/2021 12:14:45] DEBUG - Calling command '/usr/sbin/grub2-probe --target=device /boot/efi/'
[09/09/2021 12:14:45] DEBUG - Calling command 'lsblk -spnlo name /dev/vda1'
[09/09/2021 12:14:45] DEBUG - Block device: /dev/vda
[09/09/2021 12:14:45] DEBUG - ESP device number: {'major': 252, 'minor': 1}
[09/09/2021 12:14:45] DEBUG - The new EFI binary: /boot/efi/EFI/redhat/shimx64.efi
[09/09/2021 12:14:45] DEBUG - The new EFI binary [efi format]: \EFI\redhat\shimx64.efi
[09/09/2021 12:14:45] DEBUG - Calling command '/usr/sbin/efibootmgr -c -d /dev/vda -p 1 -l '\EFI\redhat\shimx64.efi' -L 'Red Hat Enterprise Linux 8''
Check the new EFI bootloader.
[09/09/2021 12:14:45] DEBUG - Calling command '/usr/sbin/efibootmgr'
BootCurrent: 0004
Timeout: 0 seconds
BootOrder: 0004,0001,0002,0000,0003
Boot0000* UiApp
Boot0001* Red Hat Enterprise Linux 8
Boot0002* UEFI Misc Device
Boot0003* EFI Internal Shell
Boot0004* Red Hat Enterprise Linux 8
[09/09/2021 12:14:45] DEBUG - Calling command '/usr/sbin/efibootmgr -v'
BootCurrent: 0004
Timeout: 0 seconds
BootOrder: 0004,0001,0002,0000,0003
Boot0000* UiApp FvVol(7cb8bdc9-f8eb-4f34-aaea-3ee4af6516a1)/FvFile(462caa21-7614-4503-836e-8ab6f4662331)
Boot0001* Red Hat Enterprise Linux 8    HD(1,GPT,714e014a-5636-47b0-a8a7-b5109bfe895c,0x800,0x12c000)/File(\EFI\redhat\shimx64.efi)
Boot0002* UEFI Misc Device      PciRoot(0x0)/Pci(0x2,0x3)/Pci(0x0,0x0)N.....YM....R,Y.
Boot0003* EFI Internal Shell    FvVol(7cb8bdc9-f8eb-4f34-aaea-3ee4af6516a1)/FvFile(7c04a583-9e3e-4f1c-ad65-e05268d0b4d1)
Boot0004* Red Hat Enterprise Linux 8    HD(1,GPT,714e014a-5636-47b0-a8a7-b5109bfe895c,0x800,0x12c000)/File(\EFI\redhat\shimx64.efi)
[09/09/2021 12:14:45] DEBUG - Calling command '/usr/sbin/grub2-probe --target=device /boot/efi/'
[09/09/2021 12:14:45] DEBUG - EFI: the next boot is not set.
[09/09/2021 12:14:45] DEBUG - Calling command '/usr/sbin/efibootmgr'
BootCurrent: 0004
Timeout: 0 seconds
BootOrder: 0004,0001,0002,0000,0003
Boot0000* UiApp
Boot0001* Red Hat Enterprise Linux 8
Boot0002* UEFI Misc Device
Boot0003* EFI Internal Shell
Boot0004* Red Hat Enterprise Linux 8
[09/09/2021 12:14:45] DEBUG - Calling command '/usr/sbin/efibootmgr -v'
BootCurrent: 0004
Timeout: 0 seconds
BootOrder: 0004,0001,0002,0000,0003
Boot0000* UiApp FvVol(7cb8bdc9-f8eb-4f34-aaea-3ee4af6516a1)/FvFile(462caa21-7614-4503-836e-8ab6f4662331)
Boot0001* Red Hat Enterprise Linux 8    HD(1,GPT,714e014a-5636-47b0-a8a7-b5109bfe895c,0x800,0x12c000)/File(\EFI\redhat\shimx64.efi)
Boot0002* UEFI Misc Device      PciRoot(0x0)/Pci(0x2,0x3)/Pci(0x0,0x0)N.....YM....R,Y.
Boot0003* EFI Internal Shell    FvVol(7cb8bdc9-f8eb-4f34-aaea-3ee4af6516a1)/FvFile(7c04a583-9e3e-4f1c-ad65-e05268d0b4d1)
Boot0004* Red Hat Enterprise Linux 8    HD(1,GPT,714e014a-5636-47b0-a8a7-b5109bfe895c,0x800,0x12c000)/File(\EFI\redhat\shimx64.efi)
[09/09/2021 12:14:45] DEBUG - Calling command '/usr/sbin/grub2-probe --target=device /boot/efi/'
[09/09/2021 12:14:45] DEBUG - EFI: the next boot is not set.
[09/09/2021 12:14:45] DEBUG - Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/convert2rhel/main.py", line 110, in main
    grub.post_ponr_set_efi_configuration()
  File "/usr/lib/python3.6/site-packages/convert2rhel/grub.py", line 590, in post_ponr_set_efi_configuration
    _replace_efi_boot_entry(efibootinfo)
  File "/usr/lib/python3.6/site-packages/convert2rhel/grub.py", line 525, in _replace_efi_boot_entry
    _remove_current_boot_entry(efibootinfo)
  File "/usr/lib/python3.6/site-packages/convert2rhel/grub.py", line 471, in _remove_current_boot_entry
    if orig_boot != efibootinfo_orig.entries[orig_boot.boot_number]:
KeyError: '0004'
```

bocekm replied:
I'd not rely on specific boot entry position (#4 in this case). Check/track what's position of the entry added by us.

pirat89 replied:
@SpyTec  @bocekm the comparison in this one case has a different meaning. (see the warning msg in the body of the condition). This one specific check is supposed to get info whether the original entry is still same or not. In case you change position, it's not original entry. Entries are not numbered strictily in order. You can have set a bootloader with entries ('0000', '0003', '0004', '0006', ..). Let's call it identifier instead of the position, this makes the additional thinking about that easier. The condition is not supposed to check a newly added entry, but the original one. That's why the numbering is strict. 

The piece of the code for the context
```python
    if orig_boot != efibootinfo_orig.entries[orig_boot.boot_number]:            
        logger.warning(                                                         
            "The EFI bootloader entry %s has been modified. Skipping the removal."
            % orig_boot.boot_number                                             
        )
```
Where:
- `orig_boot` is supposed to be the original boot entry after the conversion and bootloader update (new bootloader entry has been added already). read: bootloader entry with the same boot number after the recent changes (if exists)
- you want to compare it with the bootloader entry before the conversion.

Unless the code has been updated, I do not understand why the keyerror happens, as from the shared output, clearly the entry `0004` existed, so it cannot miss in the `efibootinfo_orig` dict.

SpyTec replied:
Thanks for the info Petr. I let it remain but just made sure that the entry exists when the check is being made

I also made a check to see if a RHEL boot entry already exists, so we shouldn't get duplicated entries

Boot order will likely be correct, so did not create any methods for verifying and correcting it

bocekm commented on convert2rhel/checks.py:
```diff
@@ -57,19 +58,43 @@ def perform_pre_ponr_checks():
-            "Conversion of UEFI systems is currently not supported, see"
-            " https://bugzilla.redhat.com/show_bug.cgi?id=1898314"
-            " for more information."
+def check_efi():
```
I've updated all user-facing messages to use UEFI. I don't see it as a big issue to have it different internally in the code. But feel free to create a subsequent PR changing the variables/functions/methods/classes/tests that use just EFI.
